buildtest/buildsystem/base.py

```diff
@@ -0,0 +1,558 @@
+           the metadata keys defined for any global recipe.
+        """
+        keys = []
+        if self.recipe:
+            keys = [x for x in self.recipe.keys() if x not in self.metadata]
+        return keys
+
+    def get(self, name):
+        """Given the name of a section (typically a build configuration name)
+           return the loaded section from self.recipe. If you need to parse
+           through just section names, use self.keys() to filter out metadata
+        """
+        return self.recipe.get(name)
+
+
+class BuilderBase:
+    """The BuilderBase is an abstract class that implements common functions for
+       any kind of builder.
+    """
+
+    def __init__(self, name, recipe_config, config_file=None):
+        """initiate a builder base. A recipe configuration (loaded) is required.
+           this can be handled easily with the BuildConfig class:
+
+           bc = BuildConfig(config_file)
+           recipe_config = bc.get("section_name")
+           builder = ScriptBuilder(recipe_config)
+
+           Parameters
+           ==========
+           name: a name for the build recipe (required)
+           recipe_config: the loaded section from the config_file for the user.
+           config_file: the pull path to the configuration file, must exist.
+        """
+        self.name = name
+        self.result = {}
+        self.build_id = None
+        self.metadata = {}
+        self.config_file = config_file
+
+        # A builder is required to define the type attribute
+        if not hasattr(self, "type"):
+            sys.exit(
+                "A builder base is required to define the 'type' as a class variable"
+            )
+
+        # The recipe must be loaded as a dictionary
+        if not isinstance(recipe_config, dict):
+            sys.exit(
+                "Please load a recipe configuration before providing to the builder."
+            )
+
+        # The type must match the type of the builder
+        self.recipe = recipe_config
+        if self.recipe.get("type") != self.type:
+            sys.exit(
+                "Mismatch in type. Builder expects %s but found %s."
+                % (self.type, self.recipe.get("type"))
+            )
+
+    def __str__(self):
+        return "[builder-%s-%s]" % (self.type, self.name)
+
+    def __repr__(self):
+        return self.__str__()
+
+    def _create_test_folders(self):
+        """Create all needed test folders on init, and add their paths
+           to self.metadata
+        """
+        testpath = os.path.expandvars(self.metadata["testpath"])
+        testdir = os.path.dirname(testpath)
+        create_dir(testdir)
+        for folder in ["run", "log"]:
+            name = "%sdir" % folder
+            self.metadata[name] = os.path.join(testdir, folder)
+            create_dir(self.metadata[name])
+
+    def _init_logger(self, to_file=True):
+        """initialize the logger. This is called at the end of prepare_run,
```

Comment:
Can we define the parameter ``to_file`` in the docstring

Reply:
Another PR.